feat(op-acceptance-tests); base; geth-reth divergence test.

scharissis opened the pull request:
**Description**

Adds a new acceptance test which checks for divergence between geth and reth nodes by comparing their state roots.


**Tests**

Ran locally against kurtosis-based isthmus and interop devnets:

**isthmus**
<img width="702" alt="Screenshot 2025-04-28 at 14 24 03" src="https://github.com/user-attachments/assets/afe63e95-de96-430e-ba78-03aaeca6facf" />

**interop**
<img width="704" alt="Screenshot 2025-04-28 at 14 30 14" src="https://github.com/user-attachments/assets/8bf3925f-94b2-4445-9eee-5ada4bd2a85f" />

Curiously, the skips here should have made the overall acceptance tests fail. I'm investigating this separately.

go test output:
```
=== RUN   TestRethGethDivergence
    reth_geth_divergence_test.go:36: INFO [04-28|14:36:17.017] Started reth-geth divergence test
=== RUN   TestRethGethDivergence/Chain_0
=== PAUSE TestRethGethDivergence/Chain_0
=== RUN   TestRethGethDivergence/Chain_1
=== PAUSE TestRethGethDivergence/Chain_1
=== CONT  TestRethGethDivergence/Chain_0
=== CONT  TestRethGethDivergence/Chain_1
=== NAME  TestRethGethDivergence
    reth_geth_divergence_test.go:118: WARN [04-28|14:36:17.017] No reth node found                       chain=0
=== NAME  TestRethGethDivergence/Chain_0
    reth_geth_divergence_test.go:65: no reth node found in chain
=== NAME  TestRethGethDivergence
    reth_geth_divergence_test.go:118: WARN [04-28|14:36:17.017] No reth node found                       chain=1
=== NAME  TestRethGethDivergence/Chain_1
    reth_geth_divergence_test.go:65: no reth node found in chain
--- PASS: TestRethGethDivergence (0.92s)
    --- SKIP: TestRethGethDivergence/Chain_0 (0.00s)
    --- SKIP: TestRethGethDivergence/Chain_1 (0.00s)
PASS
ok      github.com/ethereum-optimism/optimism/op-acceptance-tests/tests/base    3.699s
```


**Metadata**

- Fixes https://github.com/ethereum-optimism/infra/issues/270

scharissis committed:
feat(op-acceptance-tests); base; geth-reth divergence test.

Adds a new acceptance test which checks for divergence between geth and reth nodes by comparing their state roots.

scharissis commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,204 @@
+		testTimeout = 60 * time.Second
+
+		// blockCheckCount is the number of recent blocks to check for divergence
+		blockCheckCount = 5
```
Is this far back enough, or should we have a time-based check? For example, check for divergence for 2 minutes.

scharissis commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,204 @@
+	// Find geth and reth nodes
+	gethNode, rethNode, err := findGethAndRethNodes(chain, logger)
+	if err != nil {
+		t.Skip(err.Error())
```
Is it reasonable to skip here, given that not all chains will have both a geth and reth node?
Note that if we skip, the acceptance tests will fail.

bitwiseguy replied:
You could change `findGethAndRethNodes` so that it doesn't return an error if there aren't any reth nodes found. Instead you could add an extra condition here that just returns (i.e. test passes) if no reth nodes are found. Or do we explicitly want this acceptance test to skip+fail if there are no reth nodes?

scharissis replied:
Thanks. I don't think that skip+fail makes sense as it seems completely reasonable for there to be no reth nodes.
I'll change this to how you suggested while we await for more feedback here.

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	// Check each block for divergence
+	for blockNumber := startBlock; blockNumber <= endBlock; blockNumber++ {
```
Curious: why not just check for the divergence only at "endBlock" ( or `min(latestGethBlock, latestRethBlock)` )? 🤔

If any previous block diverges, it consequently ends up causing divergence for every next block.

And if a block isn't diverged, then that correlates to all of the previous blocks being un-diverged.

So, just checking the latestBlock / endBlock would tell if you divergence really ended up happening or not.

scharissis replied:
Thanks Yash - no good reason, I just forgot about this heuristic. I'll make it use the "endBlock".

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,204 @@
+		if isGethNode(nodeName) {
+			gethNode = node
+			logger.Debug("Identified geth node", "name", nodeName)
+		} else if isRethNode(nodeName) {
```
This function seems to find the last known reth node in the chain.

IMO we should return a slice of all the reth nodes because we would want to check divergence for all of them.

scharissis replied:
Agreed. Done.

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,204 @@
+	defer cancel()
+
+	// Find geth and reth nodes
+	gethNode, rethNode, err := findGethAndRethNodes(chain, logger)
```
Related to https://github.com/ethereum-optimism/optimism/pull/15585/files#r2063527237

We should iterate over `rethNodes` to subject each reth node to divergence check.

scharissis replied:
Agreed; done.

tynes commented:
Why does it need to know specifically about reth and geth? What if there is a non deterministic bug that causes a chain split in the same client? What happens when we add op-nethermind?

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,190 @@
+			}
+
+			// Get the latest block from reth
+			latestRethBlock, err := rethClient.BlockNumber(ctx)
```
Might be nit-picking but hear me out:

The current piece of code does 4 IOs with the client:
- gethClient.blockNum
- rethClient.blockNum
_Find the min block num_
- gethClient.getHeader(min block num)
- rethClient.getHeader(min block num)

I think it would be shorter and faster to just do:
- gethClient.getHeader(nil) - returns the latest block header of geth
- rethClient.getHeader(nil) - returns the latest block header of reth

Then,
```
var rethBlockToCompare, gethHeaderToCompare

// if reth is ahead of geth, get the reth's blockHeader matching geth's latest Block Header
if latestRethHeader.BlockNum < latestGethHeader.BlockNum {
    gethHeaderToCompare = gethClient.getHeader(rethHeader.BlockNum)
    rethHeaderToCompare = latestRethHeader
} else {
// if geth is ahead of reth, get the geth's blockHeader matching reth's latest Block Header
    rethHeaderToCompare = 
    rethClient.getHeader(gethHeader.BlockNum)
    gethHeaderToCompare = latestGethHeader
}

Just compare gethHeaderToCompare.StateRoot with rethHeaderToCompare.StateRoot
```

wdyt

yashvardhan-kukreja replied:
The above comment might become obsolete in favor of https://github.com/ethereum-optimism/optimism/pull/15585/files#r2066780769

scharissis replied:
Thanks. I've re-worked things as per the other comment, so I think this is now obsolete.

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/reth_geth_divergence_test.go:
```diff
@@ -0,0 +1,190 @@
+		"reth_count", len(rethNodes))
+
+	// Get client for geth node
+	gethClient, err := gethNode.GethClient()
```
Just wondering if it would be simpler and more reliable to just check for divergence across all the nodes?
Ultimately, it's a red flag in a network's health if divergence happens in any node (geth or reth).

So maybe just 
- form a slice of all nodes (and their clients) across the network 
- find the smallest latest block across all of them, and 
- check if that block's state root matches across all the clients?

wdyt

scharissis replied:
Sounds good. I've implemented this in my latest push; please take a look.
Thanks again!

scharissis commented:
> Why does it need to know specifically about reth and geth? What if there is a non deterministic bug that causes a chain split in the same client? What happens when we add op-nethermind?

@tynes - thanks for the review!

On knowing about geth/reth and op-nethermind:
I've pushing a new version in which I try to remain more client agnostic. It is my understanding that with our current interfaces `node.GethClient` will actually return _any_ compatible client.
I expect that it'll work with nethermind. But I'll ping a few more folks on this matter in particular.

On a chain split on the same client type:
I believe the updated version should address this as well, as we're checking all nodes.
However, it wouldn't explicitly call out that this was an intra-client split. Just a divergence in general.

pcw109550 reviewed (commented):
About checking the chain split, there is a utility for this, at https://github.com/ethereum-optimism/optimism/blob/968197d26ed6f25c7a9db0666275966be38df76d/devnet-sdk/testing/systest/multi_client.go#L49.

To test divergence between chains, comparing the block header hash at same height is much more safe, since block header hash's preimage contain state root and other consensus infos.

Also, note that this acceptance test is built on the legacy devnet-sdk library. We may migrate this test by using the devstack utility: https://github.com/ethereum-optimism/optimism/blob/develop/devnet-sdk/devstack/README.md

yashvardhan-kukreja commented:
> To test divergence between chains, comparing the block header hash at same height is much more safe, since block header hash's preimage contain state root and other consensus infos.

++ to that. 

I thought of this initially but was on the fence considering that the replica-healthcheck's implementation of divergence validation is backed by considering state root.

Happy to explore using Hash as well

yashvardhan-kukreja commented on op-acceptance-tests/tests/base/state_root_divergence_test.go:
```diff
@@ -0,0 +1,144 @@
+	)
+
+	return func(t systest.T, sys system.System) {
+		logger := testlog.Logger(t, log.LevelInfo)
```
Just a question regarding testlog:

Will this work fine with writing Debug logs considering its constructor mentioning `log.LevelInfo`?

scharissis replied:
No, it would need to be explicitly set to `log.LevelDebug` for debug logs.

scharissis committed:
consolidated root divergence check into existing chain test.

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -283,15 +317,19 @@ func (mc *MultiClient) verifyFollowersWithRetry(
 	}
 
-	return result, nil
+	// If we get here, we've exceeded retries
+	if len(pendingClients) == 0 {
+		return hashMismatches, nil
+	}
+	return hashMismatches, fmt.Errorf("exceeded max retries, %d clients still pending verification", len(pendingClients))
```
```suggestion
	if len(pendingClients) == 0 {
		return hashMismatches, nil
	}
	// If we get here, we've exceeded retries
	return hashMismatches, fmt.Errorf("exceeded max retries, %d clients still pending verification", len(pendingClients))
```

Are you sure that this code is reachable? It looks to me like the loop above is designed to return early when we reach the max retries? It might be cleaner to panic if we get here and there are still pending clients, if we think it should be impossible. 

Perhaps we should add a unit test for this to exercise the different code paths.

scharissis replied:
Added tests. Couldn't reach this part of the code. Yet when I read it, it seems possible. 🙃 
I'd love to add even more tests and simplify this code snippet, but I'm not familiar enough with the domain to feel confident making such a change.

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,50 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+	primaryHeader, ok := primaryItem.(*types.Header)
+	if !ok {
+		// If we're fetching a block instead of a header
+		if block, ok := primaryItem.(*types.Block); ok {
+			primaryHeader = block.Header()
```
What happens if `ok` is false?

scharissis replied:
Then we're dealing with a `types.Block` instead of a `types.Header`. If I understand correctly, `queryFn` could return either of those, so we need to handle both cases.

geoknee replied:
That's for the first `ok`. I'm asking about the `ok` on L193. Then we have something which is neither a block or header but proceed anyway?

scharissis replied:
Ah, I see! Patched.

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,50 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+			return common.Hash{}, err
+		}
+
+		// If we have a successful hash match, also check state root if possible
+		if primaryHeader != nil {
+			var followerHeader *types.Header
+			if header, ok := item.(*types.Header); ok {
+				followerHeader = header
+			} else if block, ok := item.(*types.Block); ok {
+				followerHeader = block.Header()
+			}
+
+			// If we have both headers and the block hash matches but state roots differ,
+			// return a specific error for state root divergence
+			if followerHeader != nil && hash == primaryHash && followerHeader.Root != primaryHeader.Root {
+				return hash, fmt.Errorf("state root divergence detected: primary=%s, follower=%s",
+					primaryHeader.Root.Hex(), followerHeader.Root.Hex())
+			}
```
Is it even possible for the block hashes to match while the state roots differ?

In my understanding, mismatched state roots -> mismatched block hashes, so this check doesn't really make sense. But of course matching state roots does not imply matching block hashes on its own (there could be other fields which diverge while keeping the state roots the same). It's an edge case, but still technically possible. 

So the check we want to make is if the block hashes _don't_ match, let's compare the state roots. Most likely they won't match either, but it's useful to know if they do happen to match.

yashvardhan-kukreja replied:
++ 
It's redundant to have state root check with Hash-check already in place (not vice versa though).

Rest, in my humble (non-blocking) opinion, I think it would be simpler to "only" have Hash-check, rather than the state-root-check-if-hash-check-fails. 

_Maybe because I have never encountered any divergence with matching state roots, I don't understand how the extra piece of code will help._

Rest your call 😊

yashvardhan-kukreja replied:
++ 
It's redundant to have state root check with Hash-check already in place (not vice versa though).

scharissis replied:
Without knowing the details of the hashing algorithms, I was working under the belief that there is always an (infinitesimally small) chance of a hash collision.
None the less, I have updated this as per your suggestion above.

geoknee replied:
We shouldn't worry about hash collisions here, we don't anywhere else!

scharissis committed:
added a test for verifyFollowersWithRetry

scharissis committed:
updated fetchWithConsistencyCheck hash comparisons

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+					return hash, fmt.Errorf("block hash divergence with matching state roots: primary_hash=%s, follower_hash=%s, state_root=%s",
+						primaryHash.Hex(), hash.Hex(), primaryHeader.Root.Hex())
+				} else {
+					// Both hash and state root differ
+					return hash, nil
```
Shouldn't we return an error here if things aren't matching?

scharissis replied:
Yes, we should. Good catch.

yashvardhan-kukreja commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+			return common.Hash{}, err
+		}
+
+		// If hashes don't match, also check state roots
```
We can almost assume it to be impossible for the hash collision to occur with not matching state roots

(one in 2^256 possibilities lol)

yashvardhan-kukreja commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+			return common.Hash{}, err
+		}
+
+		// If hashes don't match, also check state roots
```
I am not sure if we get anything actionable by knowing that state root matches despite divergent hashes.

As an engineer running this test, I anyway would be motivated to dig into this, the moment I observe unequal hashes irrespective of the state root.

That plus the rarity of this happening (state root matching when hashes don't), I would be inclined to just having this test based on simply comparing the hashes to be fine and credible.

ajsutton replied:
Having debugged a lot of client divergences, it would be useful to print the full block header in the error message. It's very common for the state root to differ and that requires digging a lot deeper to find what differed in the state but there are many cases where things like the base fee adjustment or gas used differs and that shows up directly in the block header and can be a useful starting point.  It can be any field in the header and is often multiple fields, so I wouldn't try to assert multiple fields, but including the full header is very handy.  It could be json which is readable but verbose or RLP which is not immediately readable but more compact and can be easily parsed.

scharissis replied:
@yashvardhan-kukreja - I've added this extra check to print a more tailored error message as per previous feedback from @geoknee .

@ajsutton , in the event the headers differ I now return an error that includes both. I haven't converted to JSON or RLP for now; will leave that for a future improvement so as to not increase the scope of this PR.

geoknee replied:
@scharissis I think in the case where the state roots match but the block hashes diverge you are still only logging out the hashes. 

Honestly I think this code could be really simple: if the block hashes do not match then throw an error with the full block headers (if they are available). Then, if a divergence occurs whoever is debugging it has all the information they need. I don't think it is necessary to have the code start singling out fields (like the state root) to see if they match. This can be done by the user.

geoknee replied:
Made a PR here https://github.com/ethereum-optimism/optimism/pull/16047 it saves 50 lines of code and I think logs out the right information when a block hash mismatch happens.

scharissis committed:
updated multiclient error messages

scharissis commented:
The acceptance tests now run per-branch and, notably, they are passing :
https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/90466/workflows/987ba0e6-eb7b-41d1-9227-eee3ace5d840?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary

yashvardhan-kukreja commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,56 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+		}
+
+		// If hashes don't match, also check state roots
+		if hash != primaryHash && primaryHeader != nil {
+			var followerHeader *types.Header
+			if header, ok := item.(*types.Header); ok {
+				followerHeader = header
+			} else if block, ok := item.(*types.Block); ok {
+				followerHeader = block.Header()
+			}
+
+			// If we have both headers and the state roots match despite hash mismatch,
+			// include this information in the error message
+			if followerHeader != nil {
+				if followerHeader.Root == primaryHeader.Root {
+					return hash, fmt.Errorf("block hash divergence with matching state roots: primary_hash=%s, follower_hash=%s, state_root=%s",
+						primaryHash.Hex(), hash.Hex(), primaryHeader.Root.Hex())
+				} else {
+					// Both hash and state root differ
+					return hash, fmt.Errorf("block hash divergence. PrimaryHeader = %+v, FollowerHeader = %+v", primaryHeader, followerHeader)
+				}
+			}
+		}
+		return hash, nil
```
```suggestion
		// no (hash)divergence situation
		if hash == primaryHash {
			return hash, nil
		}
		
		// from this point, we have divergence, so no matter what, we would return a non-nil error.
		
		// hash-divergence found but with no primary header to provide more details
		if primaryHeader == nil {
			return hash, fmt.Errorf("block hash divergence found at block %s. Unfortunately, couldn't parse more details due to the nil primaryHeader.", num.String())
		}
	
		var followerHeader *types.Header
		if header, ok := item.(*types.Header); ok {
			followerHeader = header
		} else if block, ok := item.(*types.Block); ok {
			followerHeader = block.Header()
		}
		
		// hash-divergence found with follower header details but no primary header details
		if followerHeader == nil {
			return hash, fmt.Errorf("block hash divergence found at block %s. Unfortunately, couldn't parse the nil followerHeader but the primary Header is: %+v", num.String(), primaryHeader)
		}
		
		// hash-divergence found with matching state roots
		if followerHeader.Root == primaryHeader.Root {
			return hash, fmt.Errorf("block hash divergence with matching state roots: primary_hash=%s, follower_hash=%s, state_root=%s",
						primaryHash.Hex(), hash.Hex(), primaryHeader.Root.Hex())
		}
		// hash-divergence found with diverging state roots
		return hash, fmt.Errorf("block hash divergence. PrimaryHeader = %+v, FollowerHeader = %+v", primaryHeader, followerHeader)
```

Addresses the bug of returning nil error upon `hash != primaryHash && primaryHeader == nil` and adds some more readability.

scharissis committed:
Update devnet-sdk/testing/systest/multi_client.go

Addresses the bug of returning nil error upon hash != primaryHash && primaryHeader == nil and adds some more readability.

Co-authored-by: Yashvardhan Kukreja <yashvardhan@oplabs.co>

scharissis commented:
After incorporating Yash's fix, the tests are still passing.

Also note that we'll want to rewrite this test using `op-devstack` in a follow-up PR.

yashvardhan-kukreja commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,66 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
 	if err != nil {
+		// If err contains information about matching state roots, format and return it
+		if strings.Contains(err.Error(), "block hash divergence with matching state roots") {
+			return nil, fmt.Errorf("at block #%s: %s", blockNum, err.Error())
```
```suggestion
			return nil, fmt.Errorf("at block #%s: %w", blockNum, err)
```

nit

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,66 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
 		return nil, err
 	}
 
+	// Get the state root for the primary
+	primaryHeader, ok := primaryItem.(*types.Header)
```
```suggestion
	// Get the block hash for the primary
	primaryHeader, ok := primaryItem.(*types.Header)
```

geoknee committed:
simplify handling of block hash mismatch case

log out all of the information we have

geoknee committed:
renaming and further simplification

geoknee committed:
tweak godoc to remove reference to state root

geoknee commented on devnet-sdk/testing/systest/multi_client.go:
```diff
@@ -186,15 +186,66 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
+		if block, ok := primaryItem.(*types.Block); ok {
+			primaryHeader = block.Header()
+		} else {
+			return nil, fmt.Errorf("expected header or block, got %T", primaryItem)
```
Outside of the scope of this PR, but because we are using an `interface{}` type we end up with pretty painful code like this trying to figure out which concrete type we ended up with. Using `interface{}` is mostly an antipattern and seems unnecessary here since we know we will get something which is a block header or trivially convertible to a block header. 

This could definitely be improved in follow up work cc @teddyknox.

teddyknox replied:
Agreed. It feels like a refactor could simplify this code a lot.

teddyknox replied:
Fixed in #16069, which aims to merge into this PR branch.

scharissis committed:
Update devnet-sdk/testing/systest/multi_client.go

Co-authored-by: George Knee <georgeknee@googlemail.com>

scharissis committed:
Update devnet-sdk/testing/systest/multi_client.go

Co-authored-by: Yashvardhan Kukreja <yashvardhan@oplabs.co>

teddyknox committed:
Clean up devnet-sdk fork detection logic

teddyknox committed:
Merge pull request #16047 from ethereum-optimism/gk/gethreth

simplify handling of block hash mismatch case

teddyknox commented:
I've opened https://github.com/ethereum-optimism/optimism/pull/16069 into this branch to simplify the multiclient logic @scharissis @yashvardhan-kukreja.

teddyknox committed:
Merge pull request #16069 from ethereum-optimism/teddyknox/clean-up-multi-client

Clean up devnet-sdk multiclient fork detection

scharissis commented:
I believe this is good to go now.
Do any reviewers have more comments?
@yashvardhan-kukreja , @geoknee , @teddyknox

teddyknox reviewed (approved):
LGTM
